539 add apply to property

ProkopovVitaliy opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

**Description** (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
Added possibility set `apply_to` property for product  eav attribute. 

There are two variants how to use this property in data patch generation:

1. Apply attribute for all available product types in the system:
![Screenshot 2021-04-12 at 13 53 43](https://user-images.githubusercontent.com/29776509/114385039-28656500-9b98-11eb-9016-6f2a4aaa571b.png)
If  `Apply to all products types` selected no specific product type will be specified for the attribute.

![Screenshot 2021-04-12 at 13 54 15](https://user-images.githubusercontent.com/29776509/114385299-885c0b80-9b98-11eb-8e9d-5d15ba375ca8.png)

2. Specify product types:

Unselect `Apply to all products types`, and it will be possible to specify product types:
![Screenshot 2021-04-12 at 13 55 11](https://user-images.githubusercontent.com/29776509/114385437-b6d9e680-9b98-11eb-9ac9-61925799170d.png)

![Screenshot 2021-04-12 at 13 55 29](https://user-images.githubusercontent.com/29776509/114385456-bb9e9a80-9b98-11eb-87bd-95bbff863674.png)

After that will be generated data patch with specified product types:

![Screenshot 2021-04-12 at 13 55 44](https://user-images.githubusercontent.com/29776509/114385548-d53fe200-9b98-11eb-9154-9a0335eeabbe.png)


**Fixed Issues (if relevant)**
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2-phpstorm-plugin#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2-phpstorm-plugin#539

**Questions or comments**
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

**Contribution checklist** (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with integration/functional tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

eduard13 reviewed (commented):
Hi @ProkopovVitaliy, thank you for the provided changes. But I guess providing a multiple select with available product types, would be much more user friendly. Additionally, we'll avoid entering some typos within that field.

Please let me know your thoughts.

Thank you.

ProkopovVitaliy commented:
> Hi @ProkopovVitaliy, thank you for the provided changes. But I guess providing a multiple select with available product types, would be much more user friendly. Additionally, we'll avoid entering some typos within that field.
> 
> Please let me know your thoughts.
> 
> Thank you.

Hi @eduard13 Yes you are right, this is more user friendly. But if developer have custom product type in their store, fore example `sellable`, we cannot predict such options and it can be some restriction for interact with our dialog window.So that why i added "help text block" which contain default types:
`Enter comma separated product types for which attribute will available, for example: simple,configurable,bundle,virtual,downloadable,category. Or leave this property blank if you want apply attribute for all product types.`

What about your opinion ?

DmitryFurs commented:
@ProkopovVitaliy @eduard13 Magento provides **product_types.xml** file that describes the product types. Is it possible to implement support for this file for the **apply_to** option? When typing manually, there is a chance of making a typo. There may also be third-party product types. If there was multiselect option, it would be very convenient. Thanks!

https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Downloadable/etc/product_types.xml#L9
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/ConfigurableProduct/etc/product_types.xml#L9

VitaliyBoyko commented:
@ProkopovVitaliy 
So we can add a file indexer for `product_types.xml`, and build options accordingly.

ProkopovVitaliy commented:
@eduard13 I finalized this functionality. Could you please review it?

**Main points:** 

- Added indexer for `product_types.xml`
- Added multi-select for `apply_to` property that based on parsed result form `product_types.xml` files.

**Overview:**
1. Create attribute for all product types: 
![Screenshot 2021-04-13 at 14 29 38](https://user-images.githubusercontent.com/29776509/114546757-37b2e400-9c66-11eb-9e39-8979a1d000dc.png)
![Screenshot 2021-04-13 at 14 29 52](https://user-images.githubusercontent.com/29776509/114546762-38e41100-9c66-11eb-9936-4cb6369b13a7.png)
There are no changes.

2. Create attribute  for specified product types:
![Screenshot 2021-04-13 at 14 27 33](https://user-images.githubusercontent.com/29776509/114546909-64ff9200-9c66-11eb-9ac7-d807ebfd7b9d.png)
![Screenshot 2021-04-13 at 14 28 32](https://user-images.githubusercontent.com/29776509/114546916-6630bf00-9c66-11eb-80a1-1c001d2a648e.png)
As we can see here added new multi-select field with product types. 

3. And the new one. When we decide  to create new product type we will see him in the multi-select: 
Create new product type:
![Screenshot 2021-04-13 at 14 33 08](https://user-images.githubusercontent.com/29776509/114547286-ece59c00-9c66-11eb-8965-b6a7c3fcd0a7.png)

Run reindex for plugin:
![Screenshot 2021-04-13 at 14 33 24](https://user-images.githubusercontent.com/29776509/114547316-f838c780-9c66-11eb-89f6-26d81e6852c8.png)

Open window for generation and try generate attribute for new created product type:
![Screenshot 2021-04-13 at 14 35 34](https://user-images.githubusercontent.com/29776509/114547445-21f1ee80-9c67-11eb-8369-4db245211f53.png)
![Screenshot 2021-04-13 at 14 35 50](https://user-images.githubusercontent.com/29776509/114547450-23231b80-9c67-11eb-84fe-4a8a77c0943b.png)

eduard13 commented on src/com/magento/idea/magento2plugin/actions/generation/generator/util/eav/ProductAttributeMapper.java:
```diff
@@ -71,6 +71,11 @@
                 Boolean.toString(eavEntityData.isVisibleOnFront()));
 
+        if (eavEntityData.getApplyTo() != null && !eavEntityData.getApplyTo().isEmpty()) {
+            mappedAttributes.put(EavAttribute.APPLY_TO.getAttribute(),
+                    wrapStringValueForTemplate(eavEntityData.getApplyTo()));
```
@ProkopovVitaliy, just want to bring your attention, that breaking a multiple arguments to a new line, should move all the arguments to a new empty line, like following.

No action is needed here, just wanted to add a remark, it also applies to PHP code.
```suggestion
            mappedAttributes.put(
                    EavAttribute.VISIBLE_ON_FRONT.getAttribute(),
                    Boolean.toString(eavEntityData.isVisibleOnFront())
            );
```

ProkopovVitaliy replied:
@eduard13 Thanks for recommendation, really it's looks better in your variant. I will take this into account.

Thanks!

eduard13 reviewed (commented):
No action is needed here, just wanted to add a remark, it also applies to PHP code.

eduard13 reviewed (approved):
Thank you for the feature 👍.
